### PR TITLE
Upstream config serialization and inheritance

### PR DESCRIPTION
**Pull request by jacobtomlinson**
When running schedulers and workers from the CLI in distributed there is a hidden environment variable which allows you to serialize your config for passing to other components.

This is used in `distributed.SSHCluster` and other downstream cluster managers to pass all config from the system creating the cluster to the scheduler and workers.

Being able to serialize config seems useful beyond the CLI methods, so I'm upstreaming that functionality here.

This PR updates the config initialization with a check for the `DASK_INTERNAL_INHERIT_CONFIG` environment variable. If found the `dask.config.global_config` object is initialized from this serialized config.

Also upstreamed utilities and added tests and docs.

See dask/distributed#4372 for the counterpart PR.

cc @madsbk 

- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask`

### CODE REVIEW

File: dask/config.py [hunk 1]
```diff
@@ -586,7 +588,48 @@ def check_deprecations(key: str, deprecations: dict = deprecations):
...
+    -------
+    deserialized_data: obj
+        The de-serialized data
+
+    """
+    return json.loads(base64.urlsafe_b64decode(data.encode()).decode())
+
+
 def _initialize():
+    if "DASK_INTERNAL_INHERIT_CONFIG" in os.environ:
```
Comment (martindurant) [inline dask/config.py:630]: Does this take precedence over config in files, as other environment variables do?

I would suggest the block belongs better in function `collect`, which is where other environment variables are considered, and allows you to pass in a custom env dictionary.

Comment (jacobtomlinson) [reply dask/config.py]: Thanks @martindurant that makes a lot of sense. Updated!

File: dask/tests/test_config.py [hunk 2]
```diff
@@ -489,3 +491,26 @@ def test_get_override_with():
...
+        # Take a round trip through the serialization
+        serialized = serialize({"array": {"svg": {"size": 150}}})
+        config = deserialize(serialized)
+
+        dask.config.update(dask.config.global_config, config)
+        assert dask.config.get("array.svg.size") == 150
+
+
+def test_config_inheritance():
+    os.environ["DASK_INTERNAL_INHERIT_CONFIG"] = serialize(
```
Comment (martindurant) [inline dask/tests/test_config.py:509]: You could instead/also test `collect_env` by passing a dictionary (I notice that the env parameter is not documented there).

I don't insist, happy for merge as is.

**Review by madsbk — APPROVED** [review]
LGTM

File: dask/tests/test_config.py (hunk 2 above)
Comment (jacobtomlinson) [reply dask/tests/test_config.py]: Thanks @martindurant that makes the test much neater.

File: dask/tests/test_config.py [hunk 3]
```diff
@@ -489,3 +491,22 @@ def test_get_override_with():
...
+
+        dask.config.update(dask.config.global_config, config)
+        assert dask.config.get("array.svg.size") == 150
+
+
+def test_config_inheritance():
+    config = collect_env(
+        {"DASK_INTERNAL_INHERIT_CONFIG": serialize({"array": {"svg": {"size": 150}}})}
+    )
+    assert dask.config.get("array.svg.size", config=config) == 150
```
Comment (martindurant) [inline dask/tests/test_config.py:512]: Still want to unset this value after the test?

Comment (jacobtomlinson) [reply dask/tests/test_config.py:512]: Shouldn't need to, I am no longer affecting the global Dask config.

Comment (martindurant) [reply dask/tests/test_config.py:512]: oh, `config=` - sorry...

**Pull request merged**